pandas/core/internals.py

```diff
@@ -2432,6 +2441,21 @@ def to_native_types(self, slicer=None, na_rep='', quoting=None, **kwargs):
         # we are expected to return a 2-d ndarray
         return values.reshape(1, len(values))
 
+    def concat_same_type(self, to_concat, placement=None):
+        """
+        Concatenate list of single blocks of the same type.
+        """
+        to_concat = [blk.values for blk in to_concat]
+        values = _concat._concat_categorical(to_concat, axis=self.ndim - 1)
+
+        if is_categorical_dtype(values.dtype):
```

Comment:
I *think* you could dispense with all of these if/thens and just use make_block

Reply:
Yep, I could, but I wanted to make advantage of the fact that for those case it is categorical, I don't need to re-infer that. 
Not sure how important that is for the internal ones though, as the if/else statements in make_block will not give that much of overhead (I added the `concat_same_type` machinery to be able to do that in an external Block, but that does not mean that I necessarily need to do it internally).

Reply:
its just added code and more complexity. Trying reduce code here.

Reply:
So I removed this if/else check in the categorical and datetimetz block's concat_same_type. 

But if you want, I can also completely remove all overwriting of `Block.concat_same_type` and use a generic one in the base Block type (using `_concat._concat_compat` and `make_block` instead of `np.concatenate` and `self.make_block_same_class`). And in this way only keep the ability of external block types (like geopandas) to overwrite this, and without using that ability internally. But, doing that would remove the performance improvement I mentioned above (which is not the main objective of this PR, so not too bad, but having it is still nice).